### PR TITLE
feat: single prompt block and manual-only placement option

### LINKED ISSUE #538: Reinstate manual placement
Manual placements were superseded by and converted to custom placements in #467, but they are needed as a feature. 

A user should be able to just place a prompt in an arbitrary place on the site, and disable programmatic insertion of this prompt. This is possible with custom placements, **but** until there are more than three such prompts - as the only way of disabling programmatic insertion is assigning to custom placements and then instead of using the custom placement block, inserting prompts using shortcodes. 

## Use case:

Four prompts, each should be possible to be placed ~programatically~ _manually_ on the same page.

### PR DESCRIPTION
**Pull request by dkoo**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Does a few things:

1. Reinstates "manual-only" as a placement option (not a frequency option as it was previously). Prompts with this placement option will never get programmatically inserted anywhere, so they won't be displayed unless used in a shortcode or the new Single Prompt block (see below).
2. Creates a Single Prompt block which allows you to select a single inline or manual-only prompt to be displayed where placed in a post or page. This is basically a controlled wrapper UI that renders the prompt as a `[newspack-popup]` shortcode, and will hopefully be a more user-friendly alternative to using shortcodes for this purpose.
3. Makes the help text shown alongside the Placement options in the editor sidebar more helpful, by describing the currently selected option's behavior. Also changes the label to "Position" if the prompt is an overlay. See [this comment](https://github.com/Automattic/newspack-popups/issues/538#issuecomment-833859880) for details.

Closes #538.

Note that this PR uses the `AutocompleteWithSuggestions` component [implemented here](https://github.com/Automattic/newspack-plugin/pull/952), so this PR is blocked until that PR gets deployed and the new component gets published to the NPM library. Adding an "On Hold" label to reflect this, but it can still be reviewed and tested at this point.

#### Setup and prereqs

Prior to testing, make sure you have [this branch](https://github.com/Automattic/newspack-plugin/pull/952) checked out for Newspack Plugin. Then:

1. `cd` into the `newspack-plugin/assets/components` folder.
2. Run `npm run prereleaseOnly` to build the components.
3. In the `newspack-popups` plugin folder, check out this branch and run `npm ci`.
4. Copy the entire contents of the `newspack-plugin/assets/components` folder into `newspack-popups/node_modules/newspack-components`, overwriting when prompted. This will simulate the new `AutocompleteWithSuggestions` component being available via the `newspack-components` package.
5. After the above, in the Newspack Plugin repo, check out the branch from https://github.com/Automattic/newspack-plugin/pull/960 as this is required to see the new "manual only" placement option in the Campaigns wizard.

#### Testing the manual-only placement option

1. In Newspack > Campaigns, create a new inline prompt, or edit an existing inline prompt.
2. In the Campaigns dashboard, open the popover with the Placement dropdown.
3. Confirm that there's now a "Manual Only" option in the dropdown, and if selected, that the flag next to the prompt title changes to say "Manual Only."
4. Confirm that for Overlay prompts, the "Manual Only" option doesn't appear (only available for inline prompts).
5. View the prompt in the block editor. In the Prompt Settings sidebar, confirm that "Manual Only" is an available Placement option only if the Prompt Type is set to "Inline".

#### Testing the contextual Placement help

1. View the prompt in the block editor. In the Prompt Settings sidebar, confirm that when "Prompt Type" is set to "Overlay", the Placement label changes to "Position" instead, and vice versa.
2. Change the Position/Placement option to various values, and confirm that the help text updates itself according to the selected option. [More details on what they should say here](https://github.com/Automattic/newspack-popups/issues/538#issuecomment-833859880).

#### Testing the new Single Prompt block

1. Edit a post or page.
2. In the block inserter, under the Newspack category, confirm that there's a new "Newspack Campaigns: Single Prompt" block.
3. Insert the block. The placeholder should look something like this:

Screen Shot 2021-05-12 at 2 26 43 PM

4. Save or publish the post without selecting any prompt (leaving the block in placeholder mode). In a new incognito window, confirm that the block doesn't render any markup when no prompt is selected.
5. Confirm that the suggestions that load are up to 10 of the most recent **inline** or **manual-only** prompts. Prompts with overlay, above-header, or custom placement options are not selectable with this block.
6. Type a prompt name using the autocomplete field, and confirm that it shows matching inline or manual-only prompts.
7. Select a prompt using either the autocomplete or the suggestions list. Confirm that the block preview changes to show the prompt content:

Screen Shot 2021-05-12 at 2 26 52 PM

8. Confirm that the "edit" link next to the prompt title opens that prompt in the editor, in a new tab.
9. Confirm that the prompt content shown matches the selected prompt.
10. Confirm that clicking "Clear Prompt" clears your selection and returns the block back to placeholder mode.
11. Re-select a prompt in the "Everyone" segment (or an easily matched segment) with the "Manual only" placement option and save the post.
12. View the post in a new incognito session. Confirm that the prompt is displayed exactly where you placed the block.
13. Change the prompt's placement to "inline." Refresh the post in the incognito session and confirm that the prompt is now shown twice—once where you placed the block, and again where it's programmatically inserted as an inline prompt.
14. Add another Single Prompt block to the same post, and select the same prompt. Refresh the post in the incognito session and confirm that the prompt is now shown three times—once for each instance of the block, and again for the programmatic inline placement.
15. Change the prompt's segment to something you don't match. Refresh the post and confirm that the prompt is no longer shown in any instance, because it's not in your best-priority segment.
16. Deactivate or delete the prompt. Refresh the post editor where you placed the prompt in a Single Prompt block and confirm that the block now shows a warning:

Screen Shot 2021-05-13 at 4 59 17 PM

17. Confirm on the front-end that the block with a non-existent prompt assigned doesn't render any markup.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by dkoo** [comment]
@adekbadek I could use your help in figuring out why the test I implemented in a620543 to test the block insertion in a post is failing. Is it perhaps because the block renders a shortcode, and the `renderPost` test method isn't rendering the full markup as it would be displayed in a front-end context?

**Comment by dkoo** [comment]
@thomasguillot Pinging you because I could use your help implementing WP icons to replace the placeholder Material icons used by both Custom Placement and Single Prompt blocks. I wasn't sure which WP icons would be appropriate for these. 

The block UI implemented in this PR is pretty basic, but always appreciate your feedback if you have any.

**Comment by adekbadek** [comment]
@dkoo 

> I could use your help in figuring out why the test I implemented in a620543 to test the block insertion in a post is failing.

It's because the shortcode function discards prompts which fail `Newspack_Popups_Model::is_inline` condition:

https://github.com/Automattic/newspack-popups/blob/e5c4652955a60bbdeedcb991585da92db9f472b7/includes/class-newspack-popups-inserter.php#L431-L432

**Comment by dkoo** [comment]
> It's because the shortcode function discards prompts which fail `Newspack_Popups_Model::is_inline` condition

Aha, that makes sense! Thanks for highlighting. Fixed in 1ac7369. What's strange to me is that the shortcode still rendered the popup despite the failing test, though.

**Comment by dkoo** [comment]
Removing the [On Hold] status because https://github.com/Automattic/newspack-plugin/pull/952 has landed and is queued for release. That change must still be deployed before this one for production release.

**Review by adekbadek — COMMENTED** [review]
I've found one issue with prompt preview, but it's not blocking: when the previewed prompt has a Jetpack Mailchimp signup block, the block states are all displayed at once:

image

> Removing the [On Hold] status because Automattic/newspack-plugin#952 has landed and is queued for release. That change must still be deployed before this one for production release.

I don't think it does - the `newspack-components` npm release is live and completely separate from the `newspack-plugin` release.

**Review by adekbadek — APPROVED** [review]
Meant to approve, not merely comment :D

File: includes/class-newspack-popups-api.php [hunk 1]
```diff
@@ -91,5 +117,62 @@ public static function update_settings( $request ) {
 			]
 		);
 	}
+
+	/**
+	 * Get inline prompts with the given params.
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 */
+	public static function get_inline_and_manual_prompts( $request ) {
```
[the remark this answers is not in the document]
Comment (dkoo) [reply includes/class-newspack-popups-api.php:126]: Ok, that sounds like a good thing to tackle in a separate PR.

**Review by adekbadek — APPROVED** [review]

**Comment by dkoo** [comment]
> That did not hide the elements for me, but 9f592ea did :)

Nice, thanks for the fix!

**Pull request merged**